// @sendbird/calls-react-native/lib/module/types/Platform.js
// NOTE: This file must follow the native-side convention.

/**
 * @platform iOS
 * @description AVAudioSession.RouteChangeReason {@link https://developer.apple.com/documentation/avfaudio/avaudiosession/routechangereason}
 */
export let RouteChangeReason = /*#__PURE__*/function (RouteChangeReason) {
  /// The reason is unknown.
  RouteChangeReason[RouteChangeReason["unknown"] = 0] = "unknown";
  /// A new device became available (e.g. headphones have been plugged in).
  RouteChangeReason[RouteChangeReason["newDeviceAvailable"] = 1] = "newDeviceAvailable";
  /// The old device became unavailable (e.g. headphones have been unplugged).
  RouteChangeReason[RouteChangeReason["oldDeviceUnavailable"] = 2] = "oldDeviceUnavailable";
  /// The audio category has changed (e.g. AVAudioSessionCategoryPlayback has been changed to
  /// AVAudioSessionCategoryPlayAndRecord).
  RouteChangeReason[RouteChangeReason["categoryChange"] = 3] = "categoryChange";
  /// The route has been overridden (e.g. category is AVAudioSessionCategoryPlayAndRecord and
  /// the output has been changed from the receiver, which is the default, to the speaker).
  RouteChangeReason[RouteChangeReason["override"] = 4] = "override";
  /// The device woke from sleep.
  RouteChangeReason[RouteChangeReason["wakeFromSleep"] = 5] = "wakeFromSleep";
  /// Returned when there is no route for the current category (for instance, the category is
  /// AVAudioSessionCategoryRecord but no input device is available).
  RouteChangeReason[RouteChangeReason["noSuitableRouteForCategory"] = 6] = "noSuitableRouteForCategory";
  /// Indicates that the set of input and/our output ports has not changed, but some aspect of
  /// their configuration has changed.  For example, a port's selected data source has changed.
  /// (Introduced in iOS 7.0, watchOS 2.0, tvOS 9.0).
  RouteChangeReason[RouteChangeReason["routeConfigurationChange"] = 7] = "routeConfigurationChange";
  return RouteChangeReason;
}({});

/**
 * @platform iOS
 * @description AVAudioSession.Port {@link https://developer.apple.com/documentation/avfaudio/avaudiosession/port}
 */
export let AVAudioSessionPort = /*#__PURE__*/function (AVAudioSessionPort) {
  /** input port types **/
  /// Line level input on a dock connector
  AVAudioSessionPort["lineIn"] = "lineIn";
  /// Built-in microphone on an iOS device
  AVAudioSessionPort["builtInMic"] = "builtInMic";
  /// Microphone on a wired headset.  Headset refers to an accessory that has headphone outputs paired with a microphone.
  AVAudioSessionPort["headsetMic"] = "headsetMic";
  /** output port types **/
  /// Line level output on a dock connector
  AVAudioSessionPort["lineOut"] = "lineOut";
  /// Headphone or headset output
  AVAudioSessionPort["headphones"] = "headphones";
  /// Output on a Bluetooth A2DP device
  AVAudioSessionPort["bluetoothA2DP"] = "bluetoothA2DP";
  /// The speaker you hold to your ear when on a phone call
  AVAudioSessionPort["builtInReceiver"] = "builtInReceiver";
  /// Built-in speaker on an iOS device
  AVAudioSessionPort["builtInSpeaker"] = "builtInSpeaker";
  /// Output via High-Definition Multimedia Interface
  AVAudioSessionPort["HDMI"] = "HDMI";
  /// Output on a remote Air Play device
  AVAudioSessionPort["airPlay"] = "airPlay";
  /// Output on a Bluetooth Low Energy device
  AVAudioSessionPort["bluetoothLE"] = "bluetoothLE";
  /** port types that refer to either input or output **/
  /// Input or output on a Bluetooth Hands-Free Profile device
  AVAudioSessionPort["bluetoothHFP"] = "bluetoothHFP";
  /// Input or output on a Universal Serial Bus device
  AVAudioSessionPort["usbAudio"] = "usbAudio";
  /// Input or output via Car Audio
  AVAudioSessionPort["carAudio"] = "carAudio";
  /// Input or output that does not correspond to real audio hardware
  AVAudioSessionPort["virtual"] = "virtual";
  /// Input or output connected via the PCI (Peripheral Component Interconnect) bus
  AVAudioSessionPort["PCI"] = "PCI";
  /// Input or output connected via FireWire
  AVAudioSessionPort["fireWire"] = "fireWire";
  /// Input or output connected via DisplayPort
  AVAudioSessionPort["displayPort"] = "displayPort";
  /// Input or output connected via AVB (Audio Video Bridging)
  AVAudioSessionPort["AVB"] = "AVB";
  /// Input or output connected via Thunderbolt
  AVAudioSessionPort["thunderbolt"] = "thunderbolt";
  return AVAudioSessionPort;
}({});
