// @sendbird/calls-react-native/lib/module/utils/logger.js
import { Platform } from 'react-native';
const LogLevelEnum = {
  'none': 0,
  'error': 1,
  'warning': 2,
  'info': 3
};
/** @internal **/
export const getLogger = (lv = 'none', title) => {
  let _logLevel = __DEV__ ? lv : 'none';
  let _title = title ?? `[Calls_${Platform.OS}]`;
  return {
    setTitle(title) {
      _title = title;
    },
    setLogLevel(lv) {
      if (__DEV__) _logLevel = lv;
    },
    getLogLevel() {
      return _logLevel;
    },
    error(...args) {
      if (LogLevelEnum[_logLevel] < LogLevelEnum.error) return LogLevelEnum.none;
      console.error(_title, ...args);
      return LogLevelEnum[_logLevel];
    },
    warn(...args) {
      if (LogLevelEnum[_logLevel] < LogLevelEnum.warning) return LogLevelEnum.none;
      console.warn(_title, ...args);
      return LogLevelEnum[_logLevel];
    },
    info(...args) {
      if (LogLevelEnum[_logLevel] < LogLevelEnum.info) return LogLevelEnum.none;
      console.info(_title, ...args);
      return LogLevelEnum[_logLevel];
    }
  };
};
export const Logger = getLogger();
