// @sendbird/calls-react-native/lib/commonjs/types/Call.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DirectCallUserRole = exports.DirectCallEndResult = void 0;
/** DirectCall */
let DirectCallEndResult = exports.DirectCallEndResult = /*#__PURE__*/function (DirectCallEndResult) {
  /** Default value of the EndResult. **/
  DirectCallEndResult["NONE"] = "NONE";
  /** The call has ended by either the caller or callee after successful connection. **/
  DirectCallEndResult["COMPLETED"] = "COMPLETED";
  /** The caller has canceled the call before the callee accepts or declines. **/
  DirectCallEndResult["CANCELED"] = "CANCELED";
  /** The callee has declined the call. **/
  DirectCallEndResult["DECLINED"] = "DECLINED";
  /** The call is accepted on one of the callee’s other devices. All the other devices will receive this call result. **/
  DirectCallEndResult["OTHER_DEVICE_ACCEPTED"] = "OTHER_DEVICE_ACCEPTED";
  /** Sendbird server failed to establish a media session between the caller and callee within a specific period of time. **/
  DirectCallEndResult["TIMED_OUT"] = "TIMED_OUT";
  /** Data streaming from either the caller or the callee has stopped due to a WebRTC connection issue while calling. **/
  DirectCallEndResult["CONNECTION_LOST"] = "CONNECTION_LOST";
  /** The callee has not either accepted or declined the call for a specific period of time. **/
  DirectCallEndResult["NO_ANSWER"] = "NO_ANSWER";
  /** The 'dial()' method of the call has failed. **/
  DirectCallEndResult["DIAL_FAILED"] = "DIAL_FAILED";
  /** The 'accept()' method of the call has failed. **/
  DirectCallEndResult["ACCEPT_FAILED"] = "ACCEPT_FAILED";
  /** Unknown **/
  DirectCallEndResult["UNKNOWN"] = "UNKNOWN";
  DirectCallEndResult["NOT_CONNECTED"] = "NOT_CONNECTED";
  return DirectCallEndResult;
}({});
let DirectCallUserRole = exports.DirectCallUserRole = /*#__PURE__*/function (DirectCallUserRole) {
  DirectCallUserRole["CALLER"] = "CALLER";
  DirectCallUserRole["CALLEE"] = "CALLEE";
  return DirectCallUserRole;
}({});
