// @sendbird/calls-react-native/lib/commonjs/types/Media.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VideoDevicePosition = exports.SoundType = exports.RecordingType = exports.RecordingStatus = exports.AudioDeviceType = void 0;
let SoundType = exports.SoundType = /*#__PURE__*/function (SoundType) {
  SoundType["DIALING"] = "DIALING";
  SoundType["RINGING"] = "RINGING";
  SoundType["RECONNECTING"] = "RECONNECTING";
  SoundType["RECONNECTED"] = "RECONNECTED";
  return SoundType;
}({});
let AudioDeviceType = exports.AudioDeviceType = /*#__PURE__*/function (AudioDeviceType) {
  AudioDeviceType["EARPIECE"] = "EARPIECE";
  AudioDeviceType["SPEAKERPHONE"] = "SPEAKERPHONE";
  AudioDeviceType["WIRED_HEADSET"] = "WIRED_HEADSET";
  AudioDeviceType["BLUETOOTH"] = "BLUETOOTH";
  return AudioDeviceType;
}({});
let VideoDevicePosition = exports.VideoDevicePosition = /*#__PURE__*/function (VideoDevicePosition) {
  VideoDevicePosition["FRONT"] = "FRONT";
  VideoDevicePosition["BACK"] = "BACK";
  VideoDevicePosition["UNSPECIFIED"] = "UNSPECIFIED";
  return VideoDevicePosition;
}({});
let RecordingStatus = exports.RecordingStatus = /*#__PURE__*/function (RecordingStatus) {
  RecordingStatus["NONE"] = "NONE";
  RecordingStatus["RECORDING"] = "RECORDING";
  return RecordingStatus;
}({});
let RecordingType = exports.RecordingType = /*#__PURE__*/function (RecordingType) {
  /** An option to record the video and audio of the remote user */
  RecordingType["REMOTE_AUDIO_AND_VIDEO"] = "REMOTE_AUDIO_AND_VIDEO";
  /** An option to record the audio of the remote user. */
  RecordingType["REMOTE_AUDIO_ONLY"] = "REMOTE_AUDIO_ONLY";
  /** An option to record both audios of the local and remote users. */
  RecordingType["LOCAL_REMOTE_AUDIOS"] = "LOCAL_REMOTE_AUDIOS";
  /** An option to record both audios of the local and remote users, and the video of the remote user. */
  RecordingType["LOCAL_AUDIO_REMOTE_AUDIO_AND_VIDEO"] = "LOCAL_AUDIO_REMOTE_AUDIO_AND_VIDEO";
  /** An option to record both audios of the local and remote users, and the video of the local user. */
  RecordingType["LOCAL_AUDIO_AND_VIDEO_REMOTE_AUDIO"] = "LOCAL_AUDIO_AND_VIDEO_REMOTE_AUDIO";
  return RecordingType;
}({});
