// @sendbird/calls-react-native/lib/commonjs/utils/logger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getLogger = exports.Logger = void 0;
var _reactNative = require("react-native");
const LogLevelEnum = {
  'none': 0,
  'error': 1,
  'warning': 2,
  'info': 3
};
/** @internal **/
const getLogger = (lv = 'none', title) => {
  let _logLevel = __DEV__ ? lv : 'none';
  let _title = title ?? `[Calls_${_reactNative.Platform.OS}]`;
  return {
    setTitle(title) {
      _title = title;
    },
    setLogLevel(lv) {
      if (__DEV__) _logLevel = lv;
    },
    getLogLevel() {
      return _logLevel;
    },
    error(...args) {
      if (LogLevelEnum[_logLevel] < LogLevelEnum.error) return LogLevelEnum.none;
      console.error(_title, ...args);
      return LogLevelEnum[_logLevel];
    },
    warn(...args) {
      if (LogLevelEnum[_logLevel] < LogLevelEnum.warning) return LogLevelEnum.none;
      console.warn(_title, ...args);
      return LogLevelEnum[_logLevel];
    },
    info(...args) {
      if (LogLevelEnum[_logLevel] < LogLevelEnum.info) return LogLevelEnum.none;
      console.info(_title, ...args);
      return LogLevelEnum[_logLevel];
    }
  };
};
exports.getLogger = getLogger;
const Logger = exports.Logger = getLogger();
